detectron2/export/model_convert_utils_2.py

```diff
@@ -0,0 +1,470 @@
+            print(
+                "Blob {} with type {} is not supported in generating init net,"
+                " skipped.".format(name, type(blob))
+            )
+            continue
+        add_tensor(ret, name, blob)
+
+    return ret
+
+
+def get_ws_blobs(blob_names=None):
+    """ Get blobs in 'blob_names' in the default workspace,
+        get all blobs if blob_names is None """
+    blobs = {}
+    if blob_names is None:
+        blob_names = workspace.Blobs()
+    blobs = {x: workspace.FetchBlob(x) for x in blob_names}
+
+    return blobs
+
+
+def get_device_option_cpu():
+    device_option = core.DeviceOption(caffe2_pb2.CPU)
+    return device_option
+
+
+def get_device_option_cuda(gpu_id=0):
+    device_option = caffe2_pb2.DeviceOption()
+    device_option.device_type = caffe2_pb2.CUDA
+    device_option.device_id = gpu_id
+    return device_option
+
+
+def create_input_blobs_for_net(net_def):
+    for op in net_def.op:
+        for blob_in in op.input:
+            if not workspace.HasBlob(blob_in):
+                workspace.CreateBlob(blob_in)
+
+
+def compare_model(model1_func, model2_func, test_image, check_blobs):
+    """ model_func(test_image, check_blobs)
+    """
+    cb1, cb2 = check_blobs, check_blobs
+    if isinstance(check_blobs, dict):
+        cb1 = check_blobs.keys()
+        cb2 = check_blobs.values()
+    print("Running the first model...")
+    res1 = model1_func(test_image, check_blobs)
+    print("Running the second model...")
+    res2 = model2_func(test_image, check_blobs)
+    for idx in range(len(cb1)):
+        print("Checking {} -> {}...".format(cb1[idx], cb2[idx]))
+        n1, n2 = cb1[idx], cb2[idx]
+        r1 = res1[n1] if n1 in res1 else None
+        r2 = res2[n2] if n2 in res2 else None
+        assert r1 is not None or r2 is None, "Blob {} in model1 is None".format(n1)
+        assert r2 is not None or r1 is None, "Blob {} in model2 is None".format(n2)
+        assert r1.shape == r2.shape, "Blob {} and {} shape mismatched: {} vs {}".format(
+            n1, n2, r1.shape, r2.shape
+        )
+
+        np.testing.assert_array_almost_equal(
+            r1,
+            r2,
+            decimal=3,
+            err_msg="{} and {} not matched. Max diff: {}".format(
+                n1, n2, np.amax(np.absolute(r1 - r2))
+            ),
+        )
+
+    return True
+
+
+# graph_name could not contain word 'graph'
+def save_graph(net, file_name, graph_name="net", op_only=True):
+    from caffe2.python import net_drawer
+
+    graph = None
+    ops = net.op
+    if not op_only:
+        graph = net_drawer.GetPydotGraph(ops, graph_name, rankdir="TB")
+    else:
+        graph = net_drawer.GetPydotGraphMinimal(
+            ops, graph_name, rankdir="TB", minimal_dependency=True
+        )
+
+    try:
+        graph.write_png(file_name)
+    except Exception as e:
+        print("Error when writing graph to image {}".format(e))
+
+
+def convert_model_gpu(net, init_net):
```

Comment:
@wat3rBro  is this needed or is there a way to use the device information during tracing?

Reply:
By tracing it can partly reflect the device information, if you try running conversion using `quick_schedules/mask_rcnn_R_50_FPN_inference_acc_test.yaml` (fixing DistributeFpnProposals like this PR and not using `MODEL.DEVICE cpu`), every CPU/GPU copying will become a CastOp.

However the CastOp only has information for dtype but not device (every such CastOp just says casting to fp32 in this case), it can not directly decide device by only looking at a single CastOp. Adding some extra annotation (eg. using AliasWithName) might help.

Reply:
I see, thanks for the explanation! For this PR, are you OK with us using the `convert_model_gpu()` function to "fix" the predict/init nets so that they correctly run on the GPU?
For context, this function adds the necessary CopyGPUToCPU/CopyCPUToGPU ops, and sets the CPU `device_option` for CPU-only ops.

Reply:
I see, thanks for the explanation! For this PR, are you OK with us using the `convert_model_gpu()` to "fix" the predict/init nets to correctly run on the GPU?
For context: this function adds the `CopyGPUToCPU`/`CopyCPUToGPU` ops to handle the CPU-only ops, as well as sets the CPU device_option for GPU-only ops.

Reply:
@wat3rBro do you know some better ways to convert the model on GPU than the manual filtering done in this PR?

Reply:
Discussed with ONNX team, our plan exporting CopyGPUToCPU to pytorch, we should be able to avoid the manual filtering.

Reply:
Ok, sounds good! In the mean time, let me modify this PR to remove the `convert_model_gpu()` usage, and we can merge in this change set that gets us 90% of the way towards supporting Caffe2 GPU export.
Does that sound reasonable?